### PR TITLE
Introduce an experimental CLI to help train TL programs using multiple GPUs.

### PR DESCRIPTION
**Pull request by luomai**

### CODE REVIEW

File: tensorlayer/cli/train.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import multiprocessing
+import os
+import re
+import subprocess
+import sys
+
+PORT_BASE = 10000
+
+
+def _get_gpu_ids():
```
Comment (jorgemf) [inline tensorlayer/cli/train.py]: I this code is much better and also takes into account CUDA_VISIBLE_DEVICES:
```
from tensorflow.python.client import device_lib

def get_available_gpus():
    local_device_protos = device_lib.list_local_devices()
    return [x.name for x in local_device_protos if x.device_type == 'GPU']
```

**Comment by jorgemf** [comment]
For local training with multiple GPUs TensorFlow is adding new features to estimators that replicate the model once per GPU and them combine the gradients: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/estimator/python/estimator

The problem with distributed TensorFlow is the overhead of the network communication of the model over grpc. It is very slow to encode and decode the model. There are solutions as using DRMA over MPI or Verbs but those imply to compile TensorFlow with support for them and have specialized hardware in your machines. The solution TensorFlow is doing to solve this issues locally is to replicate the model one per GPU getting rid of the network communication and grpc.

To sum up, I don't think this is a good practice, and it would be better to use the TF estimators or something similar.

File: tensorlayer/cli/train.py (hunk 1 above)
Comment (luomai) [reply tensorlayer/cli/train.py]: Hi Jorge, we were using this piece of the code to get GPUs before but it actually causes problems in tensorflow initialization so we switch to this approach temporarily.

**Comment by luomai** [comment]
@jorgemf Hi Jorge, thanks for your comment. We are aware of the estimator API as well as its mission for making distributed training transparent to users. However, a bunch of TL users would have to completely rewrite their code to fit the estimator model. Our current approach to resolve this is to provide a helper script to help them use tl.distributed to modify their program and use the train.py to bootstrap the local training. We also recommend them, in the future, to use estimator API to write their model if possible (the estimator API does not allow deep customization of the layers at this moment and we expect this would become better in the future).

We are also aware of the gRPC communication overhead. TF has some proposals to resolve using commodity hardware. However, we treat this as a separate issue with this PR. What do you think?

To sum up, the main purpose of this PR is to simply provide a helper script to start a TF_CONFIG environment locally so that existing TL programs (like the minst_distributed and inception_v3_distributed) may find easy to be setup locally. This PR does not aim to create another paradigm in parallel with the estimator.

**Comment by luomai** [comment]
@jorgemf Also, for the cluster training mode, we would recommend users to use the Google Kubeflow. We have run some internal test with the Kubeflow team. It is super easy for TL users to use Kubeflow.

**Comment by jorgemf** [comment]
@luomai I haven't done a lot of tests with more than one local GPU but in my current setup with 2 GPUs 1080ti and a small model (NasNet mobile), the improvement with 2 GPUs is only 10%. Instead of training the model in 10 days it takes 9. It means it spends the most time in the network communication rather than processing the data. That is why I think it is not a good practice and we should try to avoid it.

I am working on splitting the model as the estimators do and also integrating it with the datasets classes. I think we can provide a better API that uses the datasets, helps them with the distributed environment and also works with more than one GPU locally without doing anything else. Although we would need some way to disable the autoreplication of the model in all local GPUs for the people who already splits the model themselves.

**Comment by luomai** [comment]
@jorgemf The main purpose of this PR is to help users to automatically setup the TF_CONFIG environment for parallel TF programs locally.  We have run the inception v3 example on a Azure GPU 4 card VM using asynchronous synchronization. We observe that the GPUs utilization is around 90%. Each worker is processing their data shard at almost equivalent rate. This means that for GPU-intensive model, we could expect performance speed up more or less, though we do admit that the fact that gRPC is not very efficient for big models. As for the model convergence time, we believe this would depends on the model, data shading, as well as the synchronization approach. We regards this as a separated issue which shall be explored by the users themselves.
